relayer/listener.go

```diff
@@ -217,31 +300,76 @@ func (lstnr *Listener) ProcessLogs(ctx context.Context) error {
-				zap.String("sourceAddress", txLog.SourceAddress.String()),
+				zap.Uint64("blockNumber", block.BlockNumber),
 			)
 
-			// Messages are either catch-up messages, or live incoming messages.
-			// For live messages, we only write to the database if we're done catching up.
-			// This is because during the catch-up process, we cannot guarantee the order
-			// of live messages relative to catch-up messages, whereas we know that catch-up
-			// messages will always be ordered relative to each other.
-			err := lstnr.RouteMessage(&txLog, doneCatchingUp || txLog.IsCatchUpMessage)
-			if err != nil {
-				lstnr.logger.Error(
-					"Error relaying message",
-					zap.String("originChainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
-					zap.Error(err),
+			// Iterate over the Warp logs in two passes:
+			// The first pass extracts the information needed to relay from the log, but does not initiate relaying
+			// This is so that the number of messages to be processed can be registered with the database before
+			// any messages are processed
+			// The second pass dispatches the messages to the application relayers for processing
+			expectedMessages := make(map[database.RelayerID]uint64)
+			var msgsInfo []*parsedMessageInfo
+			for _, warpLog := range block.WarpLogs {
+				warpLogInfo, err := lstnr.NewWarpLogInfo(warpLog)
+				if err != nil {
+					lstnr.logger.Error(
+						"Failed to create warp log info",
+						zap.Error(err),
+					)
+					continue
+				}
+				msgInfo, err := lstnr.parseMessage(warpLogInfo)
+				if err != nil {
+					lstnr.logger.Error(
+						"Failed to parse message",
+						zap.String("blockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
+						zap.Error(err),
+					)
+					continue
+				}
+				if msgInfo == nil {
+					lstnr.logger.Debug(
+						"Application relayer not found. Skipping message relay.",
+						zap.String("blockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
+					)
+					continue
+				}
+				msgsInfo = append(msgsInfo, msgInfo)
+				expectedMessages[msgInfo.applicationRelayer.relayerID]++
+			}
+			lstnr.logger.Debug(
+				"Expected messages",
+				zap.Any("expectedMessages", expectedMessages),
+			)
+			for _, appRelayer := range lstnr.applicationRelayers {
+				// Prepare the each application relayer's database key with the number
+				// of expected messages. If no messages are found in the above loop, then
+				// totalMessages will be 0
+				totalMessages := expectedMessages[appRelayer.relayerID]
+				appRelayer.checkpointManager.prepareHeight(block.BlockNumber, totalMessages)
+			}
+			for _, msgInfo := range msgsInfo {
+				lstnr.logger.Info(
+					"Relaying message",
+					zap.String("sourceBlockchainID", lstnr.sourceBlockchain.GetBlockchainID().String()),
+					zap.String("warpMessageID", msgInfo.unsignedMessage.ID().String()),
 				)
-				continue
+				err := lstnr.dispatchToApplicationRelayer(msgInfo, block.BlockNumber)
```

Comment:
I agree with the suggested changes, but if you don't mind, I'd like to defer to #31. There's a decent amount of refactoring for `Listener` and `ApplicationRelayer` that we'll have to do to support that change. The current incoming `AppResponse` routing logic assumes that at most one message per `Listener` is being processed at any given time, so we'll need to update that routing logic to account for concurrent message processing. My thought right now is to update the [external handler](https://github.com/ava-labs/awm-relayer/blob/eb9162b4233c4d995dbc1086834d50e4317b3271/peers/external_handler.go#L30) to route directly to `ApplicationRelayer`s instead of indirectly through the `Listener`. That'll require some changes to both types, so would rather make all the intended changes in one pass.